addons/Trusts/Trusts.lua

```diff
@@ -0,0 +1,338 @@
+          ['2']='Mihli Aliapoh',
+          ['3']='Tenzen',
+          ['4']='Adelheid',
+          ['5']='Joachim'},
+        Japanese={
+          ['1']='ヴァレンラール',
+          ['2']='ミリ・アリアポー',
+          ['3']='テンゼン',
+          ['4']='アーデルハイト',
+          ['5']='ヨアヒム'}}[windower.ffxi.get_info().language]},
+    wait={
+      ['aftercast']=3,
+      ['retr']=1.25,
+      ['retrall']=3},}
+  settings = config.load(defaults)
+  lang = string.lower(settings.language)
+end)
+
+windower.register_event('addon command',function(...)
+  cmd = {...}
+  if S{'save','s'}[cmd[1]] then
+    save_set(cmd[2])
+  elseif S{'check','c'}[cmd[1]] then
+    check_learned()
+  else
+    call_set(cmd[1] or 'default')
+  end
+end)
+
+function save_set(set)
+  settings.sets[set] = {}
+  local trust_ind = 0
+  for i=1,5 do
+    local trust = windower.ffxi.get_party()['p'..i]
+    if trust and trust.mob.spawn_type == 14 then
+      trust_ind = trust_ind + 1
+      settings.sets[set][tostring(trust_ind)]=trusts:with('models',trust.mob.models[1])[lang]
+    end
+  end
+  settings:save('all')
+  log('set '..set..' saved.')
+end
+
+function check_lang(entity)
+  return {japanese=entity.japanese,english=entity.english}[lang];
+end
+
+function check_limit()
+  for i,v in pairs(windower.ffxi.get_key_items()) do
+    --Trust permit,Rhapsody in..
+    limit = S{2497,2499,2501}[v] and 3 or v==2884 and 4 or v==2886 and 5 or limit or 0
+  end
+  return limit;
+end
+
+function call_trust()
+  if #queue > 0 then
+    windower.send_command('input /ma "'..windower.to_shift_jis(check_lang(queue[1]))..'" <me>')
+  end
+end
+
+function check_exist()
+  local party = {} --include only trusts. --['name']=models
+  local party_ind = {} -- index of trust's name in current party. {'name1','name2',...,'name5'}
+  for i=1,5 do
+    local member = windower.ffxi.get_party()['p'..i]
+    if member then
+      if member.mob.spawn_type == 14 then
+        party[member.name]= member.mob.models[1]
+        table.insert(party_ind,member.name)
+      end
+    end
+  end
+  return {party,party_ind};
+end
+
+function call_set(set)
+  queue = {} --trusts to be cast.
+  settings = config.load()
+  local party,party_ind = unpack(check_exist())
+  local limit = check_limit() --upper limit # of calls trust in current player.
+  local time = os.clock() --window open
+
+  if set == 'random' then
+    local checked = {}
+    local others = windower.ffxi.get_party().party1_count - #party_ind - 1 --# of human in party exept <me>.
+
+    if limit == #party_ind then
+      windower.send_command('input /retr all')
+      calls = limit
+      coroutine.sleep(settings.wait.retrall)
+    else
+      calls = limit - #party_ind - others
+    end
+
+    repeat
+      local index = trusts[math.random(1,#trusts)]
+      if not table.find(checked,index.name) then
+        table.insert(checked,index.name)
+        if windower.ffxi.get_spells()[index.id] and windower.ffxi.get_spell_recasts()[index.id] == 0 then
```

Comment:
Again, cache the result of those functions in variables outside of the loop. You should always assume that every `windower.ffi.get_X` function is very expensive.